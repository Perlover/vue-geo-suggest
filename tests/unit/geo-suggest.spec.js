import { shallowMount } from '@vue/test-utils'
import GeoSuggest from '@/geo-suggest'
import gmapsStub from './fixtures/gmaps-stub'
import geocodedFixtures from './fixtures/geocoded'
import {
  rawSuggestionFixtures,
  normalizedSuggestionFixtures,
} from './fixtures/suggestions'

describe('geo-suggest', () => {
  beforeEach(() => {
    // gmapsStub.mockReset()
    jest.resetAllMocks()
  })

  it('renders nothing without slots', () => {
    const wrapper = shallowMount(GeoSuggest)
    expect(wrapper.html()).toBeFalsy()
  })

  it('renders default slot without wrappers', () => {
    const child = `<div>child component</div>`
    const wrapper = shallowMount(GeoSuggest, {
      scopedSlots: { default: child },
    })

    expect(wrapper.html()).toMatch(child)
  })

  it('provides slot props to the default slot', () => {
    shallowMount(GeoSuggest, {
      scopedSlots: {
        default(props) {
          expect(props).toMatchObject({
            loading: false,
            suggestions: [],
          })
        },
      },
    })
  })

  it('does not run before initialization', () => {
    const wrapper = shallowMount(GeoSuggest)

    wrapper.setProps({ search: '1234' })
    expect(gmapsStub.getPlacePredictions).not.toHaveBeenCalled()
  })

  it('initializes with gmaps as a prop', () => {
    const wrapper = shallowMount(GeoSuggest, {
      propsData: { googleMaps: gmapsStub },
    })

    wrapper.setProps({ search: '1234' })
    expect(gmapsStub.getPlacePredictions).toHaveBeenCalled()
  })

  it('initializes with global gmaps', () => {
    const wrapper = shallowMount(GeoSuggest)

    window.google = { maps: gmapsStub }

    wrapper.setProps({ search: '1234' })
    expect(gmapsStub.getPlacePredictions).toHaveBeenCalled()
    delete window.google
  })

  describe('suggestions', () => {
    it('does not run for short queries', () => {
      const wrapper = shallowMount(GeoSuggest, {
        propsData: {
          googleMaps: gmapsStub,
          minLength: 2,
        },
      })

      wrapper.setProps({ search: '1' })
      expect(gmapsStub.getPlacePredictions).not.toHaveBeenCalled()

      wrapper.setProps({ search: '12' })
      expect(gmapsStub.getPlacePredictions).toHaveBeenCalled()
    })
    it('accepts a debounce function', () => {
      const debounce = jest.fn(fn => fn)

      const wrapper = shallowMount(GeoSuggest, {
        propsData: {
          googleMaps: gmapsStub,
          debounce,
        },
      })

      wrapper.setProps({ search: '1234' })
      expect(debounce).toHaveBeenCalled()
    })

    it('loads when searching', () => {
      const wrapper = shallowMount(GeoSuggest, {
        propsData: {
          googleMaps: gmapsStub,
        },
      })

      gmapsStub.getPlacePredictions.mockImplementationOnce((_, callback) => {
        expect(wrapper.vm.loading).toEqual(true)
        callback([])
        expect(wrapper.vm.loading).toEqual(false)
      })

      wrapper.setProps({ search: '1234' })
    })

    it('provides normalized suggestions using options', () => {
      const search = '1234'
      const country = 'something'
      const props = {
        types: ['something'],
        location: { something: true },
        bounds: { something: true },
        radius: 5,
      }

      const wrapper = shallowMount(GeoSuggest, {
        propsData: {
          googleMaps: gmapsStub,
          country,
          ...props,
        },
      })

      gmapsStub.getPlacePredictions.mockImplementationOnce(
        (options, callback) => {
          expect(options).toMatchObject({
            ...props,
            input: search,
            sessionToken: { token: true },
            componentRestrictions: { country },
          })

          callback(rawSuggestionFixtures)
        }
      )

      let emitted = wrapper.emitted()
      expect(emitted.suggestions).toHaveLength(1)
      expect(emitted.suggestions[0][0]).toHaveLength(0)

      wrapper.setProps({ search })

      emitted = wrapper.emitted()
      expect(emitted.suggestions).toHaveLength(2)
      expect(emitted.suggestions[0][0]).toHaveLength(0)
      expect(emitted.suggestions[1][0]).toEqual(wrapper.vm.suggestions)

      expect(wrapper.vm.suggestions).toEqual(
        expect.arrayContaining(normalizedSuggestionFixtures)
      )
    })
  })

  describe('geocoded', () => {
    it('provides details about the specified place ID with options', () => {
      const placeDetailFields = ['something']
      const selectedSuggestionFixture = normalizedSuggestionFixtures[0]

      gmapsStub.getDetails.mockImplementationOnce((options, callback) => {
        expect(options).toMatchObject({
          placeId: selectedSuggestionFixture.placeId,
          fields: placeDetailFields,
          sessionToken: { token: true },
        })

        callback(geocodedFixtures, gmapsStub)
      })

      const wrapper = shallowMount(GeoSuggest, {
        propsData: {
          googleMaps: gmapsStub,
          suggestion: selectedSuggestionFixture,
          placeDetailFields,
        },
      })

      expect(wrapper.emitted().geocoded[0][0]).toMatchSnapshot()
    })

    it('geocodes an address with options', () => {
      const country = 'something'
      const props = {
        location: { something: true },
        bounds: { something: true },
      }
      const selectedSuggestionFixture = {
        ...normalizedSuggestionFixtures[0],
        placeId: undefined,
      }

      gmapsStub.geocode.mockImplementationOnce((options, callback) => {
        expect(options).toMatchObject({
          componentRestrictions: { country },
          address: selectedSuggestionFixture.description,
          ...props,
        })

        callback([geocodedFixtures], gmapsStub)
      })

      const wrapper = shallowMount(GeoSuggest, {
        propsData: {
          googleMaps: gmapsStub,
          suggestion: selectedSuggestionFixture,
          country,
          ...props,
        },
      })

      expect(wrapper.emitted().geocoded[0][0]).toMatchSnapshot()
    })

    it('emit errors', () => {
      const status = 'wrong status'

      gmapsStub.getDetails.mockImplementationOnce((_, callback) => {
        callback(geocodedFixtures, status)
      })

      const wrapper = shallowMount(GeoSuggest, {
        propsData: {
          googleMaps: gmapsStub,
          suggestion: normalizedSuggestionFixtures[0],
        },
      })

      expect(wrapper.emitted().error[0][0]).toMatchObject({ status })

      gmapsStub.geocode.mockImplementationOnce((_, callback) => {
        callback([geocodedFixtures], status)
      })

      wrapper.setProps({
        suggestion: { ...normalizedSuggestionFixtures[0], placeId: undefined },
      })

      expect(wrapper.emitted().error[1][0]).toMatchObject({ status })
    })
  })
})
