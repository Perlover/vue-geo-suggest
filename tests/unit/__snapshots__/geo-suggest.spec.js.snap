// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`geo-suggest geocoded geocodes an address with options 1`] = `
Object {
  "addressComponentMap": Object {
    "administrative_area_level_1": Object {
      "longName": "Lombardia",
      "shortName": "Lombardia",
    },
    "administrative_area_level_2": Object {
      "longName": "Città Metropolitana di Milano",
      "shortName": "MI",
    },
    "administrative_area_level_3": Object {
      "longName": "Milano",
      "shortName": "Milano",
    },
    "country": Object {
      "longName": "Italy",
      "shortName": "IT",
    },
    "locality": Object {
      "longName": "Milano",
      "shortName": "Milano",
    },
    "political": Object {
      "longName": "Italy",
      "shortName": "IT",
    },
    "postal_code": Object {
      "longName": "20124",
      "shortName": "20124",
    },
    "route": Object {
      "longName": "Piazza Duca d'Aosta",
      "shortName": "Piazza Duca d'Aosta",
    },
    "street_number": Object {
      "longName": "1",
      "shortName": "1",
    },
  },
  "description": "description",
  "formattedAddress": "Piazza Duca d'Aosta, 1, 20124 Milano MI, Italy",
  "gmaps": Object {
    "address_components": Array [
      Object {
        "long_name": "1",
        "short_name": "1",
        "types": Array [
          "street_number",
        ],
      },
      Object {
        "long_name": "Piazza Duca d'Aosta",
        "short_name": "Piazza Duca d'Aosta",
        "types": Array [
          "route",
        ],
      },
      Object {
        "long_name": "Milano",
        "short_name": "Milano",
        "types": Array [
          "locality",
          "political",
        ],
      },
      Object {
        "long_name": "Milano",
        "short_name": "Milano",
        "types": Array [
          "administrative_area_level_3",
          "political",
        ],
      },
      Object {
        "long_name": "Città Metropolitana di Milano",
        "short_name": "MI",
        "types": Array [
          "administrative_area_level_2",
          "political",
        ],
      },
      Object {
        "long_name": "Lombardia",
        "short_name": "Lombardia",
        "types": Array [
          "administrative_area_level_1",
          "political",
        ],
      },
      Object {
        "long_name": "Italy",
        "short_name": "IT",
        "types": Array [
          "country",
          "political",
        ],
      },
      Object {
        "long_name": "20124",
        "short_name": "20124",
        "types": Array [
          "postal_code",
        ],
      },
    ],
    "adr_address": <span class="street-address">Piazza Duca d&#39;Aosta, 1</span>, <span class="postal-code">20124</span> <span class="locality">Milano</span> <span class="region">MI</span>, <span class="country-name">Italy</span>,
    "formatted_address": "Piazza Duca d'Aosta, 1, 20124 Milano MI, Italy",
    "geometry": Object {
      "location": Object {
        "lat": [Function],
        "lng": [Function],
      },
      "viewport": Object {
        "east": 9.20687195000005,
        "north": 45.48827155,
        "south": 45.48323455,
        "west": 9.201300149999952,
      },
    },
    "html_attributions": Array [],
    "icon": "https://maps.gstatic.com/mapfiles/place_api/icons/train-71.png",
    "id": "450d39db5f2518103ba660860f4fe6d9f84d7d48",
    "name": "Milano Centrale Railway Station",
    "photos": Array [
      Object {
        "height": 2250,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/112161980613332308106/photos">Luca Celoria</a>,
        ],
        "width": 4000,
      },
      Object {
        "height": 3024,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/117983380951041976680/photos">Davide Alessio Valerio</a>,
        ],
        "width": 4032,
      },
      Object {
        "height": 3024,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/100213265519534715207/photos">Richard Ulloa</a>,
        ],
        "width": 4032,
      },
      Object {
        "height": 4032,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/101022040081875213348/photos">PMCP AVAC Pedro Pinto</a>,
        ],
        "width": 3024,
      },
      Object {
        "height": 2268,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/113232317324093711779/photos">claudio bolognesi</a>,
        ],
        "width": 4032,
      },
      Object {
        "height": 1520,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/117466078053165121971/photos">Lev Boonin</a>,
        ],
        "width": 10528,
      },
      Object {
        "height": 2799,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/107726458942788673867/photos">omaredza zyn</a>,
        ],
        "width": 2799,
      },
      Object {
        "height": 2340,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/101488497219228496445/photos">山崎あゆ</a>,
        ],
        "width": 4160,
      },
      Object {
        "height": 4032,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/107750181257939495317/photos">Chiara Monzani</a>,
        ],
        "width": 3024,
      },
      Object {
        "height": 2560,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/100064267148924486618/photos">Antonella Demicheli</a>,
        ],
        "width": 1920,
      },
    ],
    "place_id": "ChIJQ1j3T8XGhkcRKs-XGhSxJiI",
    "plus_code": Object {
      "compound_code": "F6P4+R5 Milan, Metropolitan City of Milan, Italy",
      "global_code": "8FQFF6P4+R5",
    },
    "rating": 4.1,
    "reference": "ChIJQ1j3T8XGhkcRKs-XGhSxJiI",
    "reviews": Array [
      Object {
        "author_name": "Elizabeth Lawrence",
        "author_url": "https://www.google.com/maps/contrib/116802196532562283162/reviews",
        "language": "en",
        "profile_photo_url": "https://lh4.googleusercontent.com/--S4HFTcucFg/AAAAAAAAAAI/AAAAAAAAAAA/ACHi3rcMymDJ2gz-10-zQk6pt-7TRXS3Mw/s128-c0x00000000-cc-rp-mo/photo.jpg",
        "rating": 1,
        "relative_time_description": "in the last week",
        "text": "Watch out for thieves posing as transport officials to help you buy Metro tickets.
Also self service machines may provide receipts first then you need to wait for actual ticket.
On Metro pickpockets were at work - nimble fingers reaching into mens trouser pockets and using cramming up to you techniques to reach into handbags bags etc.
Some local people were very helpful when caught pick pocket in act - though I had already been stung.",
        "time": 1553525913,
      },
      Object {
        "author_name": "Ali Sajan",
        "author_url": "https://www.google.com/maps/contrib/108989539556337281110/reviews",
        "language": "en",
        "profile_photo_url": "https://lh5.googleusercontent.com/-gydHfUw5Plc/AAAAAAAAAAI/AAAAAAAAFiM/Uv5asvW4Qjw/s128-c0x00000000-cc-rp-mo-ba4/photo.jpg",
        "rating": 5,
        "relative_time_description": "a week ago",
        "text": "This station is absolutely beautiful. A lot of shopping options available here, some great restaurants to eat, cafes everywhere and lovely architecture. Another thing I really liked was that there was alot of staff all over to help, all very friendly and professional.",
        "time": 1552914110,
      },
      Object {
        "author_name": "Alba Ariño",
        "author_url": "https://www.google.com/maps/contrib/116635588168418032541/reviews",
        "language": "en",
        "profile_photo_url": "https://lh5.googleusercontent.com/-RaQvc0_hxhY/AAAAAAAAAAI/AAAAAAAAAAA/ACHi3reiQWuKe9AlYBFaUDhbmqhfiu-7Hg/s128-c0x00000000-cc-rp-mo/photo.jpg",
        "rating": 1,
        "relative_time_description": "2 months ago",
        "text": "Trenord customer service is terrible. Buying tickets online a terrible error. Better go buy bus, better and cheaper.",
        "time": 1548605616,
      },
      Object {
        "author_name": "Christian Frei",
        "author_url": "https://www.google.com/maps/contrib/116632606110512037412/reviews",
        "language": "en",
        "profile_photo_url": "https://lh3.googleusercontent.com/-axT107sVfEE/AAAAAAAAAAI/AAAAAAAAANc/Wa-pswoev-I/s128-c0x00000000-cc-rp-mo-ba5/photo.jpg",
        "rating": 5,
        "relative_time_description": "3 months ago",
        "text": "Very beautiful train station! Fascinating. Could have more signs in the front of the train station (to guide you to your gate), but otherwise it a good train station!",
        "time": 1545546441,
      },
      Object {
        "author_name": "Heiner Barreto",
        "author_url": "https://www.google.com/maps/contrib/109454031956007141684/reviews",
        "language": "en",
        "profile_photo_url": "https://lh6.googleusercontent.com/-rFsRAJeaTk4/AAAAAAAAAAI/AAAAAAAAHBs/AhE_Dye76Bw/s128-c0x00000000-cc-rp-mo-ba5/photo.jpg",
        "rating": 5,
        "relative_time_description": "3 weeks ago",
        "text": "Beautiful train station you can find self-service or book direct from the station",
        "time": 1551725762,
      },
    ],
    "scope": "GOOGLE",
    "types": Array [
      "train_station",
      "transit_station",
      "point_of_interest",
      "establishment",
    ],
    "url": "https://maps.google.com/?cid=2460848946308173610",
    "user_ratings_total": 2396,
    "utc_offset": 120,
    "vicinity": "Piazza Duca d'Aosta, 1, Milano",
    "website": "http://www.milanocentrale.it/",
  },
  "location": Object {
    "lat": 45.4870123,
    "lng": 9.205478999999968,
  },
  "matchedSubstrings": "matched",
  "name": "Milano Centrale Railway Station",
  "normalizedAddress": Object {
    "city": "Milano",
    "countryIso2": "IT",
    "countryName": "Italy",
    "postalCode": "20124",
    "region": "MI",
    "streetAddress1": "Piazza Duca d'Aosta, 1",
    "streetAddress2": "Milano Centrale Railway Station",
  },
  "placeId": undefined,
  "region": "MI",
  "types": Array [
    "train_station",
    "transit_station",
    "point_of_interest",
    "establishment",
  ],
}
`;

exports[`geo-suggest geocoded provides details about the specified place ID with options 1`] = `
Object {
  "addressComponentMap": Object {
    "administrative_area_level_1": Object {
      "longName": "Lombardia",
      "shortName": "Lombardia",
    },
    "administrative_area_level_2": Object {
      "longName": "Città Metropolitana di Milano",
      "shortName": "MI",
    },
    "administrative_area_level_3": Object {
      "longName": "Milano",
      "shortName": "Milano",
    },
    "country": Object {
      "longName": "Italy",
      "shortName": "IT",
    },
    "locality": Object {
      "longName": "Milano",
      "shortName": "Milano",
    },
    "political": Object {
      "longName": "Italy",
      "shortName": "IT",
    },
    "postal_code": Object {
      "longName": "20124",
      "shortName": "20124",
    },
    "route": Object {
      "longName": "Piazza Duca d'Aosta",
      "shortName": "Piazza Duca d'Aosta",
    },
    "street_number": Object {
      "longName": "1",
      "shortName": "1",
    },
  },
  "description": "description",
  "formattedAddress": "Piazza Duca d'Aosta, 1, 20124 Milano MI, Italy",
  "gmaps": Object {
    "address_components": Array [
      Object {
        "long_name": "1",
        "short_name": "1",
        "types": Array [
          "street_number",
        ],
      },
      Object {
        "long_name": "Piazza Duca d'Aosta",
        "short_name": "Piazza Duca d'Aosta",
        "types": Array [
          "route",
        ],
      },
      Object {
        "long_name": "Milano",
        "short_name": "Milano",
        "types": Array [
          "locality",
          "political",
        ],
      },
      Object {
        "long_name": "Milano",
        "short_name": "Milano",
        "types": Array [
          "administrative_area_level_3",
          "political",
        ],
      },
      Object {
        "long_name": "Città Metropolitana di Milano",
        "short_name": "MI",
        "types": Array [
          "administrative_area_level_2",
          "political",
        ],
      },
      Object {
        "long_name": "Lombardia",
        "short_name": "Lombardia",
        "types": Array [
          "administrative_area_level_1",
          "political",
        ],
      },
      Object {
        "long_name": "Italy",
        "short_name": "IT",
        "types": Array [
          "country",
          "political",
        ],
      },
      Object {
        "long_name": "20124",
        "short_name": "20124",
        "types": Array [
          "postal_code",
        ],
      },
    ],
    "adr_address": <span class="street-address">Piazza Duca d&#39;Aosta, 1</span>, <span class="postal-code">20124</span> <span class="locality">Milano</span> <span class="region">MI</span>, <span class="country-name">Italy</span>,
    "formatted_address": "Piazza Duca d'Aosta, 1, 20124 Milano MI, Italy",
    "geometry": Object {
      "location": Object {
        "lat": [Function],
        "lng": [Function],
      },
      "viewport": Object {
        "east": 9.20687195000005,
        "north": 45.48827155,
        "south": 45.48323455,
        "west": 9.201300149999952,
      },
    },
    "html_attributions": Array [],
    "icon": "https://maps.gstatic.com/mapfiles/place_api/icons/train-71.png",
    "id": "450d39db5f2518103ba660860f4fe6d9f84d7d48",
    "name": "Milano Centrale Railway Station",
    "photos": Array [
      Object {
        "height": 2250,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/112161980613332308106/photos">Luca Celoria</a>,
        ],
        "width": 4000,
      },
      Object {
        "height": 3024,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/117983380951041976680/photos">Davide Alessio Valerio</a>,
        ],
        "width": 4032,
      },
      Object {
        "height": 3024,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/100213265519534715207/photos">Richard Ulloa</a>,
        ],
        "width": 4032,
      },
      Object {
        "height": 4032,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/101022040081875213348/photos">PMCP AVAC Pedro Pinto</a>,
        ],
        "width": 3024,
      },
      Object {
        "height": 2268,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/113232317324093711779/photos">claudio bolognesi</a>,
        ],
        "width": 4032,
      },
      Object {
        "height": 1520,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/117466078053165121971/photos">Lev Boonin</a>,
        ],
        "width": 10528,
      },
      Object {
        "height": 2799,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/107726458942788673867/photos">omaredza zyn</a>,
        ],
        "width": 2799,
      },
      Object {
        "height": 2340,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/101488497219228496445/photos">山崎あゆ</a>,
        ],
        "width": 4160,
      },
      Object {
        "height": 4032,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/107750181257939495317/photos">Chiara Monzani</a>,
        ],
        "width": 3024,
      },
      Object {
        "height": 2560,
        "html_attributions": Array [
          <a href="https://maps.google.com/maps/contrib/100064267148924486618/photos">Antonella Demicheli</a>,
        ],
        "width": 1920,
      },
    ],
    "place_id": "ChIJQ1j3T8XGhkcRKs-XGhSxJiI",
    "plus_code": Object {
      "compound_code": "F6P4+R5 Milan, Metropolitan City of Milan, Italy",
      "global_code": "8FQFF6P4+R5",
    },
    "rating": 4.1,
    "reference": "ChIJQ1j3T8XGhkcRKs-XGhSxJiI",
    "reviews": Array [
      Object {
        "author_name": "Elizabeth Lawrence",
        "author_url": "https://www.google.com/maps/contrib/116802196532562283162/reviews",
        "language": "en",
        "profile_photo_url": "https://lh4.googleusercontent.com/--S4HFTcucFg/AAAAAAAAAAI/AAAAAAAAAAA/ACHi3rcMymDJ2gz-10-zQk6pt-7TRXS3Mw/s128-c0x00000000-cc-rp-mo/photo.jpg",
        "rating": 1,
        "relative_time_description": "in the last week",
        "text": "Watch out for thieves posing as transport officials to help you buy Metro tickets.
Also self service machines may provide receipts first then you need to wait for actual ticket.
On Metro pickpockets were at work - nimble fingers reaching into mens trouser pockets and using cramming up to you techniques to reach into handbags bags etc.
Some local people were very helpful when caught pick pocket in act - though I had already been stung.",
        "time": 1553525913,
      },
      Object {
        "author_name": "Ali Sajan",
        "author_url": "https://www.google.com/maps/contrib/108989539556337281110/reviews",
        "language": "en",
        "profile_photo_url": "https://lh5.googleusercontent.com/-gydHfUw5Plc/AAAAAAAAAAI/AAAAAAAAFiM/Uv5asvW4Qjw/s128-c0x00000000-cc-rp-mo-ba4/photo.jpg",
        "rating": 5,
        "relative_time_description": "a week ago",
        "text": "This station is absolutely beautiful. A lot of shopping options available here, some great restaurants to eat, cafes everywhere and lovely architecture. Another thing I really liked was that there was alot of staff all over to help, all very friendly and professional.",
        "time": 1552914110,
      },
      Object {
        "author_name": "Alba Ariño",
        "author_url": "https://www.google.com/maps/contrib/116635588168418032541/reviews",
        "language": "en",
        "profile_photo_url": "https://lh5.googleusercontent.com/-RaQvc0_hxhY/AAAAAAAAAAI/AAAAAAAAAAA/ACHi3reiQWuKe9AlYBFaUDhbmqhfiu-7Hg/s128-c0x00000000-cc-rp-mo/photo.jpg",
        "rating": 1,
        "relative_time_description": "2 months ago",
        "text": "Trenord customer service is terrible. Buying tickets online a terrible error. Better go buy bus, better and cheaper.",
        "time": 1548605616,
      },
      Object {
        "author_name": "Christian Frei",
        "author_url": "https://www.google.com/maps/contrib/116632606110512037412/reviews",
        "language": "en",
        "profile_photo_url": "https://lh3.googleusercontent.com/-axT107sVfEE/AAAAAAAAAAI/AAAAAAAAANc/Wa-pswoev-I/s128-c0x00000000-cc-rp-mo-ba5/photo.jpg",
        "rating": 5,
        "relative_time_description": "3 months ago",
        "text": "Very beautiful train station! Fascinating. Could have more signs in the front of the train station (to guide you to your gate), but otherwise it a good train station!",
        "time": 1545546441,
      },
      Object {
        "author_name": "Heiner Barreto",
        "author_url": "https://www.google.com/maps/contrib/109454031956007141684/reviews",
        "language": "en",
        "profile_photo_url": "https://lh6.googleusercontent.com/-rFsRAJeaTk4/AAAAAAAAAAI/AAAAAAAAHBs/AhE_Dye76Bw/s128-c0x00000000-cc-rp-mo-ba5/photo.jpg",
        "rating": 5,
        "relative_time_description": "3 weeks ago",
        "text": "Beautiful train station you can find self-service or book direct from the station",
        "time": 1551725762,
      },
    ],
    "scope": "GOOGLE",
    "types": Array [
      "train_station",
      "transit_station",
      "point_of_interest",
      "establishment",
    ],
    "url": "https://maps.google.com/?cid=2460848946308173610",
    "user_ratings_total": 2396,
    "utc_offset": 120,
    "vicinity": "Piazza Duca d'Aosta, 1, Milano",
    "website": "http://www.milanocentrale.it/",
  },
  "location": Object {
    "lat": 45.4870123,
    "lng": 9.205478999999968,
  },
  "matchedSubstrings": "matched",
  "name": "Milano Centrale Railway Station",
  "normalizedAddress": Object {
    "city": "Milano",
    "countryIso2": "IT",
    "countryName": "Italy",
    "postalCode": "20124",
    "region": "MI",
    "streetAddress1": "Piazza Duca d'Aosta, 1",
    "streetAddress2": "Milano Centrale Railway Station",
  },
  "placeId": "place",
  "region": "MI",
  "types": Array [
    "train_station",
    "transit_station",
    "point_of_interest",
    "establishment",
  ],
}
`;
